### DISCUSSION TITLE
Override default file based cache location

### DISCUSSION THREAD

**Discussion by jordan-edrop** [vercel/next.js#72855] [Ideas]
### Goals

1. Improve experience when self hosting

### Background

We are self hosting a Docker image that is deployed to multiple clusters, but this results in two independent caches that is difficult to keep in sync.

This is addressed in the docs [here](https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath)
> You can configure the Next.js cache location if you want to persist cached pages and data to durable storage, or share the cache across multiple containers or instances of your Next.js application.

However you have to create a [custom cache implementation](https://nextjs.org/docs/app/building-your-application/deploying#configuring-caching). The docs are a little disingenuous with the example, and in reality the [NextJS cache](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/lib/incremental-cache/file-system-cache.ts) is full of internal logic that if you don't replicate will miss out on features.

The current workaround is to copy this file, change the file system path to point to the shared storage. However this is just an internal file to NextJS and is frequently updated with no actual documentation, as it is not treated as a public facing API

### Proposal

Add the ability via `next.config.js` to specify the path used by the [file-system-cache](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L33C22-L33C37)

**Comment by waynes-labs** [comment]
This is greatly needed. I wanted to mount a docker volume at a file path to use as the cache. For that to work, all the cache entries need to be in one folder, ideally a configurable one or even simply `.next/cache`. Currently I am using the following cache handler to consolidate the cache, but an official way from NextJS would be ideal.

```javascript
// cache-handler.mjs
// A Cache Handler that makes NextJS store all cache entries in its own .next/cache folder.
import RawFileSystemCache from "next/dist/server/lib/incremental-cache/file-system-cache.js";
import path from "node:path";

const FileSystemCache =
  RawFileSystemCache.default?.default ?? RawFileSystemCache.default;

export default class CacheHandler extends FileSystemCache {
  constructor(options) {
    if (!options.serverDistDir) {
      throw new Error("Expected serverDistDir");
    }
    super({
      ...options,
      serverDistDir: path.join(options.serverDistDir, "../cache/server"),
    });
  }
}
```
